src/metabase/lib/metadata/calculation.cljc

```diff
@@ -610,23 +607,31 @@
   and would thus be returned by this function.
 
   Columns from joins, expressions, and implicitly joinable columns are included automatically by default;
-  see [[VisibleColumnsOptions]] for the options for disabling these columns."
+  see `::visible-columns.options` for the options for disabling these columns."
   ([query]
-   (visible-columns query (lib.util/query-stage query -1)))
+   (visible-columns query -1))
 
-  ([query x]
-   (visible-columns query -1 x))
+  ([query stage-number]
+   ;; apparently the FE sometimes accidentally calls this with a `nil` stage number -- see #31366 -- in that case just
+   ;; return an empty vector. The FE only has access to this arity, so we don't need to do the check below.
+   ;; See [[metabase.lib.metadata.calculation-test/visible-columns-test-2]]
+   (if (and query stage-number)
+     (visible-columns query stage-number nil)
+     []))
 
-  ([query stage-number x]
-   (visible-columns query stage-number x nil))
-
-  ([query          :- ::lib.schema/query
+  ([query          :- [:maybe ::lib.schema/query]
     stage-number   :- :int
-    x
     options        :- [:maybe ::visible-columns.options]]
-   (let [options (merge (default-visible-columns-options) options)]
-     (lib.metadata.cache/with-cached-value query (cache-key ::visible-columns query stage-number x options)
-       (visible-columns-method query stage-number x options)))))
+   (let [options (-> (merge (default-visible-columns-options) options)
+                     ;; minor caching optimization: only keep the keys that are relevant to `visible-columns` --
+                     ;; ignore any other ones. As a bonus, this means undocumented options keys that aren't part of
+                     ;; the schema will effectively be ignored.
+                     (select-keys visible-columns-options-keys))]
+     (lib.metadata.cache/with-cached-value query (cache-key ::visible-columns query stage-number nil options)
```

Comment:
Is there any value in `nil` here? It keeps things parallel with `::returned-columns` I guess, but that parallel is only seen by humans. These cache keys are produced and consumed in exactly one place AFAIK - the `with-cached-value` call.

Reply:
yeah honestly I just didn't want to update `cache-key` or the underlying `lib.metadata.cache/cache-key` to support an additional arity